File: web/src/tools/implementations/geo.ts
import { z } from 'zod';
import { type LocalTool } from '../types';

const BASE_URL = '/api/v1/tools/geo';

interface GeoFeature {
  type: string;
  properties: Record<string, any>;
  geometry: {
    type: string;
    coordinates: any;
  };
  bbox?: number[];
}

interface GeocodeResponse {
  type: string;
  features: GeoFeature[];
  query?: {
    text?: string;
    parsed?: Record<string, any>;
    lat?: number;
    lon?: number;
  };
}

interface RoutingResponse {
  type: string;
  features: GeoFeature[];
  properties?: Record<string, any>;
}

interface StaticMapResponse {
  url: string;
  width: number;
  height: number;
}

interface PlacesResponse {
  type: string;
  features: GeoFeature[];
}

interface GeoToolResponse {
  success: boolean;
  data?: GeocodeResponse | RoutingResponse | StaticMapResponse | PlacesResponse;
  error?: string;
}

export const geoTool: LocalTool = {
  name: 'geoapify',
  description: 'Access Geoapify geocoding, routing, places, and map visualization services. IMPORTANT: If you need the user\'s current location and don\'t have coordinates yet, FIRST use the "browser_geolocation" tool to get their current position, then use this tool. Supports: 1) Forward geocoding, 2) Reverse geocoding, 3) Routing, 4) Static map, 5) Places (find nearby POIs). For location-based queries, get coordinates with browser_geolocation first.',
  parameters: z.object({
    action: z.enum(['geocode_search', 'geocode_reverse', 'routing', 'static_map', 'places']).describe('Action to perform: geocode_search, geocode_reverse, routing, static_map, places (find nearby POIs like supermarkets, restaurants).'),

    // Geocode search parameters
    text: z.string().optional().describe('Address or place name to search for (required for geocode_search).'),

    // Geocode reverse parameters
    lat: z.number().optional().describe('Latitude for reverse geocoding or routing waypoint.'),
    lon: z.number().optional().describe('Longitude for reverse geocoding or routing waypoint.'),

    // Routing parameters
    waypoints: z.array(z.object({
      lat: z.number().describe('Latitude of waypoint'),
      lon: z.number().describe('Longitude of waypoint'),
    })).optional().describe('Array of waypoints for routing (minimum 2 points required).'),
    mode: z.enum(['drive', 'truck', 'bicycle', 'walk', 'transit', 'approximated_transit']).optional().describe('Transportation mode for routing (default: drive).'),

    // Common parameters
    lang: z.string().optional().describe('Language code for results (e.g., "en", "de", "fr").'),
    limit: z.number().min(1).max(20).optional().describe('Maximum number of results for geocode_search (default: 10).'),
    filter: z.string().optional().describe('Filter results by location (e.g., "countrycode:us,gb").'),
    bias: z.string().optional().describe('Bias results towards a location (e.g., "countrycode:us").'),
    type: z.string().optional().describe('Type filter for reverse geocoding (e.g., "street", "city", "country").'),
    details: z.array(z.string()).optional().describe('Additional details to include in routing response.'),

    // Static map parameters
    center: z.object({
      lat: z.number(),
      lon: z.number(),
    }).optional().describe('Center point of the map (required for static_map if no markers).'),
    zoom: z.number().min(1).max(20).optional().describe('Zoom level for static map (1-20, default: auto-calculated from markers).'),
    width: z.number().min(100).max(1200).optional().describe('Map image width in pixels (default: 600).'),
    height: z.number().min(100).max(1200).optional().describe('Map image height in pixels (default: 400).'),
    style: z.enum(['osm-bright', 'osm-bright-grey', 'osm-bright-smooth', 'klokantech-basic', 'osm-liberty', 'maptiler-3d', 'toner', 'toner-grey', 'positron', 'positron-blue', 'positron-red', 'dark-matter', 'dark-matter-brown', 'dark-matter-dark-grey', 'dark-matter-dark-purple', 'dark-matter-purple-roads', 'dark-matter-yellow-roads']).optional().describe('Map style (default: osm-bright-smooth).'),
    markers: z.array(z.object({
      lat: z.number(),
      lon: z.number(),
      type: z.enum(['awesome', 'material']).optional(),
      color: z.string().optional().describe('Hex color (e.g., "#bb3f73")'),
      size: z.enum(['small', 'medium', 'large', 'x-large']).optional(),
      icon: z.string().optional().describe('Icon name (e.g., "paw", "tree", "home")'),
      icontype: z.enum(['awesome', 'material']).optional(),
      text: z.string().optional().describe('Text to display on marker'),
    })).optional().describe('Markers to display on the map.'),

    // Places parameters
    categories: z.array(z.string()).optional().describe('Categories of places to find (e.g., ["commercial.supermarket", "catering.restaurant"]).'),
    name: z.string().optional().describe('Filter places by name.'),
  }),
  execute: async ({ action, text, lat, lon, waypoints, mode, lang, limit, filter, bias, type, details, center, zoom, width, height, style, markers, categories, name }) => {
    try {
      const params: Record<string, any> = {};

      if (action === 'geocode_search') {
        if (!text) throw new Error('Text parameter is required for geocode_search action');
        params.text = text;
        if (lang) params.lang = lang;
        if (limit) params.limit = limit;
        if (filter) params.filter = filter;
        if (bias) params.bias = bias;
      }

      if (action === 'geocode_reverse') {
        if (lat === undefined || lon === undefined) {
          throw new Error('Lat and lon parameters are required for geocode_reverse action');
        }
        params.lat = lat;
        params.lon = lon;
        if (lang) params.lang = lang;
        if (type) params.type = type;
      }

      if (action === 'routing') {
        if (!waypoints || waypoints.length < 2) {
          throw new Error('At least 2 waypoints are required for routing action');
        }
        params.waypoints = waypoints;
        if (mode) params.mode = mode;
        if (details) params.details = details;
      }

      if (action === 'static_map') {
        if (center) params.center = center;
        if (zoom) params.zoom = zoom;
        if (width) params.width = width;
        if (height) params.height = height;
        if (style) params.style = style;
        if (markers) params.markers = markers;
      }

      if (action === 'places') {
        if (!filter && !bias) {
          throw new Error('Either filter (e.g. rect or circle) or bias is required for places action');
        }
        if (categories) params.categories = categories;
        if (filter) params.filter = filter;
        if (bias) params.bias = bias;
        if (limit) params.limit = limit;
        if (lang) params.lang = lang;
        if (name) params.name = name;
      }

      const response = await fetch(BASE_URL, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          action,
          params,
        }),
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`API Error (${response.status}): ${errorText}`);
      }

      const result = await response.json() as GeoToolResponse;

      if (!result.success) {
        throw new Error(result.error || 'Unknown error from Geoapify API');
      }

      // Format results based on action type
      if (action === 'static_map') {
        // Static map returns a URL directly
        const mapData = result.data as StaticMapResponse;
        return {
          type: 'static_map',
          url: mapData.url,
          width: mapData.width,
          height: mapData.height,
          message: `Map image generated. Display the image using this URL: ${mapData.url}`,
        };
      }

      if (result.data && 'features' in result.data) {
        const data = result.data as GeocodeResponse | RoutingResponse;

        if (action === 'routing') {
          // Format routing response
          const routingData = data as RoutingResponse;
          const feature = routingData.features[0];

          return {
            type: 'routing',
            distance: feature?.properties?.distance,
            distance_units: feature?.properties?.distance_units,
            time: feature?.properties?.time,
            mode: feature?.properties?.mode,
            waypoints: feature?.properties?.waypoints,
            legs: feature?.properties?.legs,
            geometry: feature?.geometry,
          };
        } else if (action === 'places') {
          // Format places response
          const placesData = data as PlacesResponse;
          const formattedResults = placesData.features.map((feature: GeoFeature) => ({
            name: feature.properties.name,
            address: feature.properties.formatted,
            categories: feature.properties.categories,
            distance: feature.properties.distance,
            lat: feature.properties.lat,
            lon: feature.properties.lon,
            place_id: feature.properties.place_id,
            details: feature.properties.details,
            opening_hours: feature.properties.opening_hours,
            website: feature.properties.website,
            phone: feature.properties.contact?.phone,
          }));

          return {
            type: 'places',
            count: formattedResults.length,
            results: formattedResults,
          };
        } else {
          // Format geocoding response
          const geocodeData = data as GeocodeResponse;
          const formattedResults = geocodeData.features.map((feature: GeoFeature) => ({
            formatted: feature.properties.formatted,
            address_line1: feature.properties.address_line1,
            address_line2: feature.properties.address_line2,
            country: feature.properties.country,
            country_code: feature.properties.country_code,
            state: feature.properties.state,
            city: feature.properties.city,
            postcode: feature.properties.postcode,
            street: feature.properties.street,
            housenumber: feature.properties.housenumber,
            lat: feature.properties.lat,
            lon: feature.properties.lon,
            result_type: feature.properties.result_type,
            category: feature.properties.category,
            timezone: feature.properties.timezone,
            plus_code: feature.properties.plus_code,
            rank: feature.properties.rank,
          }));

          return {
            type: action,
            query: geocodeData.query,
            count: formattedResults.length,
            results: formattedResults,
          };
        }
      }

      return result.data;

    } catch (error) {
      throw new Error(`Geoapify Tool Error: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  },
};

File: internal/handler/geo.go
package handler

import (
	"encoding/json"
	"llm-router/internal/model"
	"llm-router/internal/tools/geo"
	"net/http"

	"go.uber.org/zap"
)

type GeoToolRequest struct {
	Action string                 `json:"action"`
	Params map[string]interface{} `json:"params"`
}

type GeoToolResponse struct {
	Success bool        `json:"success"`
	Data    interface{} `json:"data,omitempty"`
	Error   string      `json:"error,omitempty"`
}

func HandleGeoTool(w http.ResponseWriter, r *http.Request, cfg *model.Config) {
	if cfg.GeoapifyAPIKey == "" {
		cfg.Logger.Warn("Geoapify API key not configured")
		respondWithError(w, "Geoapify API key not configured", http.StatusServiceUnavailable)
		return
	}

	var req GeoToolRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		cfg.Logger.Error("Failed to decode Geo tool request", zap.Error(err))
		respondWithError(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	client := geo.NewClient(cfg.GeoapifyAPIKey)

	var result interface{}
	var err error

	switch req.Action {
	case "geocode_search":
		geocodeReq := parseGeocodeSearchRequest(req.Params)
		result, err = client.GeocodeSearch(geocodeReq)

	case "geocode_reverse":
		reverseReq := parseGeocodeReverseRequest(req.Params)
		result, err = client.GeocodeReverse(reverseReq)

	case "routing":
		routingReq := parseRoutingRequest(req.Params)
		result, err = client.Routing(routingReq)

	case "static_map":
		staticMapReq := parseStaticMapRequest(req.Params)
		mapURL, err := client.StaticMap(staticMapReq)
		if err == nil {
			result = map[string]interface{}{
				"url":    mapURL,
				"width":  staticMapReq.Width,
				"height": staticMapReq.Height,
			}
		}

	case "places":
		placesReq := parsePlacesRequest(req.Params)
		result, err = client.Places(placesReq)

	default:
		respondWithError(w, "Unknown action: "+req.Action, http.StatusBadRequest)
		return
	}

	if err != nil {
		cfg.Logger.Error("Geoapify API request failed", zap.String("action", req.Action), zap.Error(err))
		respondWithError(w, err.Error(), http.StatusInternalServerError)
		return
	}

	respondWithJSON(w, GeoToolResponse{
		Success: true,
		Data:    result,
	})
}

func parseGeocodeSearchRequest(params map[string]interface{}) geo.GeocodeSearchRequest {
	req := geo.GeocodeSearchRequest{}

	if v, ok := params["text"].(string); ok {
		req.Text = v
	}
	if v, ok := params["lang"].(string); ok {
		req.Lang = v
	}
	if v, ok := params["limit"].(float64); ok {
		req.Limit = int(v)
	}
	if v, ok := params["filter"].(string); ok {
		req.Filter = v
	}
	if v, ok := params["bias"].(string); ok {
		req.Bias = v
	}

	return req
}

func parseGeocodeReverseRequest(params map[string]interface{}) geo.GeocodeReverseRequest {
	req := geo.GeocodeReverseRequest{}

	if v, ok := params["lat"].(float64); ok {
		req.Lat = v
	}
	if v, ok := params["lon"].(float64); ok {
		req.Lon = v
	}
	if v, ok := params["lang"].(string); ok {
		req.Lang = v
	}
	if v, ok := params["type"].(string); ok {
		req.Type = v
	}

	return req
}

func parseRoutingRequest(params map[string]interface{}) geo.RoutingRequest {
	req := geo.RoutingRequest{}

	if v, ok := params["waypoints"].([]interface{}); ok {
		for _, wp := range v {
			if wpMap, ok := wp.(map[string]interface{}); ok {
				waypoint := geo.Waypoint{}
				if lat, ok := wpMap["lat"].(float64); ok {
					waypoint.Lat = lat
				}
				if lon, ok := wpMap["lon"].(float64); ok {
					waypoint.Lon = lon
				}
				req.Waypoints = append(req.Waypoints, waypoint)
			}
		}
	}
	if v, ok := params["mode"].(string); ok {
		req.Mode = v
	}
	if v, ok := params["details"].([]interface{}); ok {
		for _, detail := range v {
			if s, ok := detail.(string); ok {
				req.Details = append(req.Details, s)
			}
		}
	}

	return req
}

func parseStaticMapRequest(params map[string]interface{}) geo.StaticMapRequest {
	req := geo.StaticMapRequest{}

	if v, ok := params["style"].(string); ok {
		req.Style = v
	}
	if v, ok := params["width"].(float64); ok {
		req.Width = int(v)
	}
	if v, ok := params["height"].(float64); ok {
		req.Height = int(v)
	}
	if v, ok := params["zoom"].(float64); ok {
		req.Zoom = v
	}

	// Parse center
	if centerMap, ok := params["center"].(map[string]interface{}); ok {
		center := &geo.Waypoint{}
		if lat, ok := centerMap["lat"].(float64); ok {
			center.Lat = lat
		}
		if lon, ok := centerMap["lon"].(float64); ok {
			center.Lon = lon
		}
		req.Center = center
	}

	// Parse markers
	if markersArray, ok := params["markers"].([]interface{}); ok {
		for _, m := range markersArray {
			if markerMap, ok := m.(map[string]interface{}); ok {
				marker := geo.MapMarker{}
				if lat, ok := markerMap["lat"].(float64); ok {
					marker.Lat = lat
				}
				if lon, ok := markerMap["lon"].(float64); ok {
					marker.Lon = lon
				}
				if typ, ok := markerMap["type"].(string); ok {
					marker.Type = typ
				}
				if color, ok := markerMap["color"].(string); ok {
					marker.Color = color
				}
				if size, ok := markerMap["size"].(string); ok {
					marker.Size = size
				}
				if icon, ok := markerMap["icon"].(string); ok {
					marker.Icon = icon
				}
				if iconType, ok := markerMap["icontype"].(string); ok {
					marker.IconType = iconType
				}
				if text, ok := markerMap["text"].(string); ok {
					marker.Text = text
				}
				req.Markers = append(req.Markers, marker)
			}
		}
	}

	if v, ok := params["area"].(string); ok {
		req.Area = v
	}

	return req
}

func parsePlacesRequest(params map[string]interface{}) geo.PlacesRequest {
	req := geo.PlacesRequest{}

	if v, ok := params["categories"].([]interface{}); ok {
		for _, cat := range v {
			if s, ok := cat.(string); ok {
				req.Categories = append(req.Categories, s)
			}
		}
	}
	if v, ok := params["filter"].(string); ok {
		req.Filter = v
	}
	if v, ok := params["bias"].(string); ok {
		req.Bias = v
	}
	if v, ok := params["limit"].(float64); ok {
		req.Limit = int(v)
	}
	if v, ok := params["lang"].(string); ok {
		req.Lang = v
	}
	if v, ok := params["name"].(string); ok {
		req.Name = v
	}

	return req
}
